source/mothership-design.md

```diff
@@ -0,0 +1,771 @@
+   compute fabric.
+
+ - *Debug packet*: An addressed item sent over DebugLink (UART) connection in
+   the Tinsel backend (`P_Debug_Pkt_t`, see the Debugging section).
+
+ - *Thread*: POSIX thread running under the Mothership process (x86-land). NB:
+   Not a "thread" in the compute fabric.
+
+ - *Mothership*: Occasionally, I refer to the Mothership as a class (or
+   object/instance), and a process. This is legacy - documentation is written
+   to take people from the past to the present after all. I'll try to be
+   explicit wherever I use this term.
+
+Figure 1 shows the class structure diagram for the proposed Mothership design.
+
+![Mothership class structure diagram](images/mothership_data_structure.pdf)
+
+# Threads and Queues: Producer-Consumer
+To support these features, a Producer-Consumer approach is used by the
+Mothership. Figure 2 shows a schematic of how the Mothership process employs
+this pattern using POSIX threads. Each thread has access to the Mothership
+object (of which there is only one per Mothership process), in which queues and
+mutexes for the producer-consumer pattern are stored. Consumer threads have
+exactly one spinner[^spinners], which is either a:
+
+ - Fast Spinners: These spin aggressively on the consumer queue, with no
+   delay between checks, ever.
+
+ - Slow Spinners: Once an event triggers the spinner, and has been resolved,
+   the next event is immediately checked for. If there is no next event, the
+   spinner delays for a brief period before checking again, and delays after
+   each check until an event triggers the spinner.
+
+[^spinners]: Spinner: Event loop, where events are items (packets or messages)
+    in the consumer queue.
+
+The discrepancy between spinner types is to encourage context to prioritise
+threads with fast spinners. The threads are:
+
+ - `main`: The root thread, which spawns the other threads below and waits for
+   them to exit. All threads are running for the duration of the Mothership
+   process (excepting process startup and teardown).
+
+ - `MPIInputBroker`: Responsible for filtering MPI messages into either the
+   Mothership object C&C queue (`MPICncQueue`), or the Application-handling
+   queue (`MPIApplicationQueue`). This "splitting" is needed because, if a
+   poorly-written supervisor[^malicious] floods with MPI messages, it will not
+   be possible for the operator to stop the application in reasonable time,
+   potentially compromising the running of other applications on the compute
+   fabric. It calls `CommonBase::MPISpinner`. This is a fast spinner.
+
+ - `MPICncResolver`: Responsible for draining the `MPICncQueue` queue, and
+   enacting those messages in sequence as appropriate. This is a slow spinner.
+
+ - `MPIApplicationResolver`: As above, but for the `MPIApplicationQueue`
+   queue. This is a slow spinner. These messages will either be:
+
+   - Converted into packets for the compute fabric, and placed in the
```

Comment:
That's odd, they are for me:

![mothership-docx-bullets](https://user-images.githubusercontent.com/12502071/81272112-f02edb80-9044-11ea-9438-bd6472d28ac6.png)

Any idea how we might resolve this?

Reply:
Interesting, which version of word are you viewing in and what dependencies might I be missing to cause these two problems? 

As the issue is not letting me upload my docx, I have emailed it to you.

Reply:
I've opened the file you sent me, and the images display in calligra 3.2.0 and libreoffice 6.3.6.2 30 on a machine without any of the source images. The image does not display in Word on Southampton's SVE service (no idea what version of Word it is - don't know how to check).

I've tried building to ODT, with the same result.

No idea how to diagnose this issue - do you know why Word might not display an image? I'm pretty sure the image is embedded and not linked (because I can open the file successfully on a machine without the source images).

Reply:
So, I have installed Libreoffice 6.4.4.2 on Windows 10 and it has the same bullet point issue as Word: 
![image](https://user-images.githubusercontent.com/6616536/82737257-b558c300-9d27-11ea-9020-90c3b0fe46d6.png)

Reply:
I've noticed that the indentation for nested list is inconsistent in this document - I've addressed this in `c471442`. Could you try opening the file that I have emailed to you on your machine, and check the nested bullets?

Reply:
Yep, that appears to have fixed the indentation

![image](https://user-images.githubusercontent.com/6616536/82894750-fb41a100-9f4a-11ea-82ce-6eb27e48e29b.png)